pkg/globalinit/plugin_discovery_source_update.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package globalinit
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/vmware-tanzu/tanzu-cli/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/lastversion"
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/config"
+)
+
+// This global initializer checks if the last executed CLI version is < 1.3.0.
+// If so it tries to update the plugin discovery source if needed
+
+func init() {
+	RegisterInitializer("Plugin Discovery Source Updater", triggerForPluginDiscoverySourceUpdater, updatePluginDiscoverySource)
+}
+
+func triggerForPluginDiscoverySourceUpdater() bool {
+	// If the last executed CLI version is < 1.3.0, we need to update the discovery image source
+	return lastversion.GetLastExecutedCLIVersion() == lastversion.OlderThan1_3_0
+}
+
+// updatePluginDiscoverySource updates the plugin discovery source to point CLI to
+// new `projects.packages.broadcom.com` and also clears the EULA acceptance state
+// stored in the CLI config file. So that EULA prompt is shown to the user again.
+func updatePluginDiscoverySource(outStream io.Writer) error {
+	ds, err := config.GetCLIDiscoverySource("default")
+	if err != nil {
+		return err
+	}
+
+	if ds == nil || ds.OCI == nil || ds.OCI.Image != constants.TanzuCLIOldCentralPluginDiscoveryImage {
+		// User must have manually updated discovery source. So do not overwrite that change.
+		return nil
+	}
+
+	// Considering we are updating the discovery source to point to new registry,
+	// User must be prompted to access the EULA again. So, let's first unset EULA status
+	err = config.SetEULAStatus(config.EULAStatusUnset)
+	if err != nil {
+		return err
+	}
+
+	// Update the `default` discovery source to point to the new central plugin discovery image
+	ds.OCI.Image = constants.TanzuCLIDefaultCentralPluginDiscoveryImage
+	fmt.Fprintf(outStream, "Updating default plugin discovery source to %q...\n", constants.TanzuCLIDefaultCentralPluginDiscoveryImage)
+	return config.SetCLIDiscoverySource(*ds)
```

Comment:
Do you think we should refresh the cache when we change the discovery source here?
We can do this with `discovery.RefreshDiscoveryDatabaseForSource(ds, discovery.WithForceRefresh())`.

I think it would be better, although the only scenario I can think of is if the broadcom repo has different data than the vmware repo, then the cache would be invalid; but still, that will only affect shell completion since other cases will notice that the cache needs to be refreshed so it is not dramatic.

Reply:
One more point, Should we fetch plugin inventory before user accepts EULA?

Reply:
> We can add that as well but do we need to use `discovery.WithForceRefresh()`? If we try to do the refresh without force option, it will only refresh the database if the digest is mismatched right?

“Refresh” means “check the digest” even if the cache TTL has not expired but it will not download the db if the digest matches. If we don’t force a refresh, then the TTL will be respected, and the digest may not be checked. (To force a db download there’s “invalidate” ). My own chosen terms, so they may not be as clear as they should be. 

> One more point, Should we fetch plugin inventory before user accepts EULA?

Hmm good point.  Ok so forget my comment. The cache will get fixed by itself eventually and there’s no dangerous side effect, so let’s keep it like you had. Maybe with a comment?

Reply:
💯 . Added a comment to explain this.